lib/device.js

```diff
@@ -1,206 +1,173 @@
+  }).bind(this));
+
+  this.mqtt.on('offline', (function() {
+    this.emit('offline');
+  }).bind(this));
+
+  this.mqtt.on('error', (function(err) {
+    this.emit('error', err);
+  }).bind(this));
+}
+
+util.inherits(Device, EventEmitter);
+
+/**
+ * Returns whether or not the device is connected.
+ */
+Device.prototype.isConnected = function() {
+  return this.mqtt.client ? this.mqtt.client.connected : false;
+};
+
+/**
+ * Occurs whenever the underlying mqtt client connects or reconnects.
+ */
+Device.prototype.handleConnect = function() {
+  debug('Subscribing to MQTT topic: ' + this.commandTopic);
+  this.mqtt.client.subscribe(this.commandTopic);
+};
+
+/**
+ * Connects to the Losant platform.
+ */
+Device.prototype.connect = function(callback) {
+  this.mqtt.connect(this.options, callback);
+};
+
+/**
+ * Disconnects the underlying mqtt connection.
+ */
+Device.prototype.disconnect = function(callback) {
+  debug('Disconnect MQTT client.');
+  if(this.mqtt.client) {
+    this.mqtt.client.end(callback);
+  }
+  else {
+    callback();
+  }
 };
+
+/**
+ * Handles a message sent from the Losant broker.
+ * topic: the mqtt topic.
+ * message: the message as a Buffer.
+ * Returns null if message is invalid. Returns the parsed JavaScript
+ * object if the message is valid.
+ */
+Device.prototype.handleMessage = function(topic, message) {
+
+  // This message is not intended for this device.
+  if(topic !== this.commandTopic) {
+    return null;
+  }
+
+  message = this.parseMessage(message);
+  if(!message) {
+    return null;
+  }
+
+  debug('Received MQTT message on topic: ' + topic);
+  debug(JSON.stringify(message));
+
+  this.emit('command', message);
+
+  return message;
+};
+
+/**
+ * Parses an incoming mqtt message from the Losant broker..
+ * Returns an object or null.
+ * message: the incoming Buffer to parse and validate.
+ */
+Device.prototype.parseMessage = function(message) {
+
+  // Attempt to parse the message.
+  try { message = EJSON.parse(message.toString()); }
+  catch(e) { return null; }
+
+  // The message has to be something.
+  if(!message) {
+    return null;
+  }
+
+  return message;
+};
+
+/**
+ * Sends device state to Losant.
+ * deviceId: the ID of the device to report state for.
+ * state: the state to send. Typically an object with { attribute: value }
+ * time (optional): the time at which the state occurred (defaults to now)
+ * cb: the callback handling the response.
+ */
+Device.prototype.sendState = function(state, time, callback) {
+
+  if(typeof(time) === 'function') {
+    callback = time;
+    time = null;
+  }
+
+  time = time || new Date();
+  var payload = { time: time, data: state };
+
+  debug('Publishing state:');
+  debug(EJSON.stringify(payload));
+  debug('Topic: ' + this.stateTopic);
+
+  // May not be connected. Skip publish if not.
+  if(this.mqtt.client) {
+    this.mqtt.client.publish(this.stateTopic, EJSON.stringify(payload), callback);
+  }
+  else {
+    if(callback) { callback(null); }
```

Comment:
the callback from the actual publish is an err/success cb with 2 args, and I would think that this would qualify as an err case